dev/TEAL.md

```diff
@@ -0,0 +1,250 @@
+
+The TEAL bytecode plus the length of any Args must add up to less than 1000 bytes (consensus parameter LogicSigMaxSize). Each TEAL op has an associated cost estimate and the program cost estimate must total less than 20000 (consensus parameter LogicSigMaxCost). Most ops have an estimated cost of 1, but a few slow crypto ops are much higher.
+
+
+
+## Constants
+
+Constants are loaded into the environment into storage separate from the stack. They can then be pushed onto the stack by referring to the type and index. This makes for efficient re-use of byte constants used for account addresses, etc.
+
+The assembler will hide most of this, allowing simple use of `int 1234` and `byte 0xcafed00d`. These constants will automatically get assembled into int and byte pages of constants, de-duplicated, and operations to load them from constant storage space inserted.
+
+Constants are loaded into the environment by two opcodes, `intcblock` and `bytecblock`. Both of these use [proto-buf style variable length unsigned int](https://developers.google.com/protocol-buffers/docs/encoding#varint), reproduced [here](#varuint). The `intcblock` opcode is followed by a varuint specifying the length of the array and then that number of varuint. The `bytecblock` opcode is followed by a varuint array length then that number of pairs of (varuint, bytes) length prefixed byte strings. This should efficiently load 32 and 64 byte constants which will be common as addresses, hashes, and signatures.
+
+Constants are pushed onto the stack by `intc`, `intc_[0123]`, `bytec`, and `bytec_[0123]`. The assembler will typically handle converting `int N` or `byte N` into the appropriate constant-offset opcode or opcode followed by index byte.
+
+## Operations
+
+Most operations work with only one type of argument, uint64 or bytes, and panic if the wrong type value is on the stack.
+The instruction set was designed to execute calculator-like expressions.
+What might be a one line expression with various parenthesized clauses should be efficiently representable in TEAL.
+
+Looping is not possible, by design, to ensure predictably fast execution.
+There is a branch instruction (`bnz`, branch if not zero) which allows forward branching only so that some code may be skipped.
+
+Many programs need only a few dozen instructions. The instruction set has some optimization built in. `intc`, `bytec`, and `arg` take an immediate value byte, making a 2-byte op to load a value onto the stack, but they also have single byte versions for loading values 0, 1, 2, and 3. Any program will benefit from having a few common values loaded with a smaller one byte opcode, and many programs won't even need more than those 4 values. Cryptographic hashes and `ed25519verify` are single byte opcodes with powerful libraries behind them. These operations still take more time than other ops (and this is reflected in the cost of each op and the cost limit of a program) but are efficient in compiled code space.
```

Comment:
I feel like 'immediate' is a standard term when referring to opcode languages.
`bnz` takes an int16 two-byte immediate branch offset (currently limited to positive int15 values so that we could maybe extend it to signed uint16 in the future).
`gtxn` takes two immediates, the group index of the transaction it's looking at and the field to read from that transaction and push to the stack.

Reply:
How about rewording line 61 slightly, as follows:

opcodes typically take one byte to specify, but some opcodes have immediate argument making them longer. The instruction set has some optimization built in. For example the opcodes `intc`, `bytec`, and `arg` take an immediate value byte (denoting an index), making a 2-byte op to load a value onto the stack. But there are also single byte versions for loading the values indexed by 0, 1, 2, and 3. Any program will benefit from having a few common values loaded with a smaller one byte opcode, and many programs won't even need more than those 4 values.
Cryptographic hashes and `ed25519verify` are single byte opcodes with powerful libraries behind them. These operations still take more time than other ops (and this is reflected in the cost of each op and the cost limit of a program) but are efficient in compiled code space.
Many useful programs need only a few dozen instructions.